Reuse QgsPostgresConn to prevent deadlock

tomtor opened the pull request:
QgsPostgresProviderConnection::executeSqlPrivate() has a nested
call to itself which may cause a deadlock when the connectionpool is
exhausted:

Two active threads can acquire a Postgres connection at the start of
executeSqlPrivate() and end up both waiting to acquire the second
connection.

## Description

[Replace this with some text explaining the rationale and details about this pull request]

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

tomtor committed:
Reuse QgsPostgresConn to prevent deadlock

QgsPostgresProviderConnection::executeSqlPrivate() has a nested
call to itself which may cause a deadlock when the connectionpool is
exhausted:

Two active threads can acquire a Postgres connection at the start of
executeSqlPrivate() and end up both waiting to acquire the second
connection.

m-kuhn reviewed (approved):
Good catch.
I guess it's not that easy to come up with a unit test for that

tomtor commented:
> Good catch.

@m-kuhn Note that I also had an alternative: d8928312a3faf4442c028f4158b8b79f31011d99
but that keeps the connection open. Not sure what is better.

elpaso commented:
Looks good to me. I should have foreseen this problem.

m-kuhn commented:
If I'm not wrong, we can have multiple postgres provider connections open in parallel for a long time, therefore keeping it open in there might deplete the connection pool rather quickly. I prefer this approach here.

Sidenote: If you want to go one step further, implementing RAII for the connection (to make `releaseConnection` exception proof) that'd be nice, but it's totally understandable if this is out of scope for this PR.

tomtor committed:
Create QgsPoolPostgresConn for RAII usage.

Wraps acquireConnection() and releaseConnection() from a QgsPostgresConnPool.

This can be used for creating std::shared_ptr<PoolPostgresConn>.

tomtor commented:
>  If you want to go one step further, implementing RAII for the connection (to make `releaseConnection` exception proof) that'd be nice, but it's totally understandable if this is out of scope for this PR.

@m-kuhn See https://github.com/qgis/QGIS/pull/40097/commits/1956130d8cabe0d3f12c11ef0d8c6c8cde4ad6b6

as a start. Should be used at more places.

m-kuhn commented on src/providers/postgres/qgspostgresconn.h:
```diff
@@ -188,6 +188,15 @@ class QgsPostgresResult
 
 };
 
+class PoolPostgresConn
```
Can you add documentation that explains the rationale for this class?

tomtor replied:
done

tomtor replied:
(and renamed to Pgs...)

tomtor committed:
Use std::make_shared()
